### DISCUSSION TITLE
Translation to other languages

### DISCUSSION THREAD

**Discussion by sencagri** [immense/Remotely#223] [Ideas]
Hello,
Is this project ready for localization for other languages ?

**Comment by stelas** [comment]
I would also appreciate it if someone could lay the foundation for adding translations into other languages. See also #238.
Any chances for that, @lucent-sea?

**Reply by bitbound** [reply]
Thanks for the feedback.

This is doable.  I've implemented localization in another app a long time ago using the Multi-Lingual App Toolkit.  It doesn't appear to be actively developed anymore, but the process should be the same.  Just more manual steps without the toolkit.

The biggest challenge is lack of time.  I only have a few hours a week to work on Remotely (family and job that pays the bills comes first), so I have to prioritize carefully.

Currently, my priorities are:
- Get the server rewrite finalized and fully released
- Finish the Mac agent (x64 processor architecture)
- Investigate changes needed for M1/Apple Silicon (might be as simple as changing the target processor architecture)
- Refactoring the quick support client to include sticky sessions and passwords for one-time unattended access
- Finding/implementing cross-platform video encoding.

Considering current velocity, finishing the above is far enough down the road that it's hard to guess what will take priority next.

But this will certainly be considered.  :)

**Reply by Creogenesis** [reply]
Can you at least tell me how to translate the client part of the project - the client downloads page, the client program. I can voluntarily support the translation. This is routine work which requires time. I understand that there is no time. But give me direction, what tools and way to do it in my fork?

**Reply by flavioxx** [reply]
I also wanted to know which files the texts are in to do the translation.

**Comment by alencarwell** [comment]
I also wanted to know which files the texts are in to do the translation.

**Comment by Martin511** [comment]
Same here.
@bitbound If you just tell us how to do it, we will do a lot of work without stealing your time ;-)

**Reply by bitbound** [reply]
It's a pretty involved process.  I don't know the best frameworks and tooling to use these days, but the gist of it is...

- Create a service that looks up strings in a dictionary.
- The dictionary gets populated with strings for a given language .
  - The target language would either be the system's default language, or you could add the ability select your desired language somewhere.
- Remove all the literal strings everywhere in the UI.
- Create a dictionary key for each place where you've removed text.
  - If the text has a dynamic value, it'd be represented by {0}, {1}, etc.  as would be used in `string.Format`.
- Bind the UI's text to a data field that gets populated by looking in the dictionary with the specific key.

Here's an example of the `strings.json` file for `fr-fr` language for the Azure Pipelines agent.  There's a strings.json file for each supported language.

![image](https://github.com/immense/Remotely/assets/20995508/46e141de-41a3-4dc6-b726-b94b023cdda9)


This PR has an example of this process:  https://github.com/immense/Remotely/pull/362/files#diff-91e417abd7402a617431565f1189376bd65c13e6348230cae3bd131ccba9100e


Unfortunately, I never got around to reviewing it, and the PR is now stale.  It would need to be caught up with `master` branch.  Even then, there isn't anyone available who has the time to effectively review it.

I did post a request for maintainers, but no one has expressed interested yet.

**Reply by Martin511** [reply]
😄 Ok, I give up.
I don´t understand a word.
Good old times when there were en-gb.ini files to copy and translate in the desired language 😅

**Reply by bitbound** [reply]
> Good old times when there were en-gb.ini files to copy and translate in the desired language ??

Well, the problem is, I didn't develop it with translation in mind.  If I had, then it would be similar to this.  You'd just copy a JSON file and translate.

**Reply by flavioxx** [reply]
> It's a pretty involved process. I don't know the best frameworks and tooling to use these days, but the gist of it is...
> 
> * Create a service that looks up strings in a dictionary.
> * The dictionary gets populated with strings for a given language .
>   
>   * The target language would either be the system's default language, or you could add the ability select your desired language somewhere.
> * Remove all the literal strings everywhere in the UI.
> * Create a dictionary key for each place where you've removed text.
>   
>   * If the text has a dynamic value, it'd be represented by {0}, {1}, etc.  as would be used in `string.Format`.
> * Bind the UI's text to a data field that gets populated by looking in the dictionary with the specific key.
> 
> Here's an example of the `strings.json` file for `fr-fr` language for the Azure Pipelines agent. There's a strings.json file for each supported language.
> 
> ![image](https://private-user-images.githubusercontent.com/20995508/344823462-46e141de-41a3-4dc6-b726-b94b023cdda9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTk4NzQyMTMsIm5iZiI6MTcxOTg3MzkxMywicGF0aCI6Ii8yMDk5NTUwOC8zNDQ4MjM0NjItNDZlMTQxZGUtNDFhMy00ZGM2LWI3MjYtYjk0YjAyM2NkZGE5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzAxVDIyNDUxM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY1ZmM0NTYzMmNkYzNmZmE2NzYxYzI1N2JhNjE3NzhmYjIwYWRjYTkzODNhMzU2ZTJmYTVjYjY1ZTUxYzVjOTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.zVzn63gzQ-I7S08dQT9NrjUMkWveC_MGxXuWDG_j2no)
> 
> This PR has an example of this process: https://github.com/immense/Remotely/pull/362/files#diff-91e417abd7402a617431565f1189376bd65c13e6348230cae3bd131ccba9100e
> 
> Unfortunately, I never got around to reviewing it, and the PR is now stale. It would need to be caught up with `master` branch. Even then, there isn't anyone available who has the time to effectively review it.
> 
> I did post a request for maintainers, but no one has expressed interested yet.

And when it is already installed by docker? What are the file paths please?

**Reply by Martin511** [reply]
Of course, I understand that. 
You did a good job anyway.
I just wanted to try to help but this is too complex for me. Not enough skills for that.
Keep going on on english ;-)
